### PR TITLE
Choose proper Conda environment path when `TRITON_DALI_SKIP_DOWNLOAD=ON`

### PR DESCRIPTION
**Pull request by szalpal**
Depending on how `TRITON_DALI_SKIP_DOWNLOAD` option in the cmake is set, the DALI Backend has to choose the conda environment path accordingly. Therefore, this PR adds an `if` to do that.

Also, since there needs to be a way to pass the `TRITON_DALI_SKIP_DOWNLOAD` option from CMake to the C++, this PR adds a mechanism to transfer the variables this way. It's a two-step process - firstly using the CMake's `configure_file` and then adjusting these variables to C++ syntax and type system.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -20,6 +20,8 @@
 # CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 configure_file(libtriton_dali.ldscript libtriton_dali.ldscript COPYONLY)
+configure_file(${tritondalibackend_SOURCE_DIR}/src/utils/purgatory.h.in
+               ${tritondalibackend_SOURCE_DIR}/src/utils/purgatory.h @ONLY)
```
Comment (mzient) [inline src/CMakeLists.txt:24]: This name is quite baffling. Wouldn't something as simple as "config" do?
```suggestion
configure_file(${tritondalibackend_SOURCE_DIR}/src/utils/config.h.in
               ${tritondalibackend_SOURCE_DIR}/src/utils/config.h @ONLY)
```

File: src/utils/cmake_to_cpp.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef DALI_BACKEND_UTILS_CMAKE_TO_CPP_H_
+#define DALI_BACKEND_UTILS_CMAKE_TO_CPP_H_
+
+#include "src/utils/purgatory.h"
```
Comment (mzient) [inline src/utils/cmake_to_cpp.h:26]: I don't quite get why this needs to be in a separate file - in principle, this file could be configured.

File: src/utils/cmake_to_cpp.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+#define STRINGIFY2(x) #x
+#define STRINGIFY(x) STRINGIFY2(x)
+
+static constexpr bool SKIP_DALI_DOWNLOAD = TRITON_DALI_SKIP_DOWNLOAD;
+static const char* CONDA_ENVIRONMENT_PATH = STRINGIFY(CONDA_ENVIRONMENT);
```
Comment (mzient) [inline src/utils/cmake_to_cpp.h:32]: Where does CONDA_ENVIRONMENT come from? Why are the mechanics for `TRITON_DALI_SKIP_DOWNLOAD` and `CONDA_ENVIRONMENT` different?

File: src/utils/cmake_to_cpp.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+#ifndef DALI_BACKEND_UTILS_CMAKE_TO_CPP_H_
+#define DALI_BACKEND_UTILS_CMAKE_TO_CPP_H_
+
+#include "src/utils/purgatory.h"
+
+#define STRINGIFY2(x) #x
+#define STRINGIFY(x) STRINGIFY2(x)
+
+static constexpr bool SKIP_DALI_DOWNLOAD = TRITON_DALI_SKIP_DOWNLOAD;
```
Comment (mzient) [inline src/utils/cmake_to_cpp.h:31]: Apart from potentially causing linker issues, I see no reason to duplicate a definition as a constant.

File: src/utils/purgatory.h.in [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef DALI_BACKEND_UTILS_PURGATORY_H_
+#define DALI_BACKEND_UTILS_PURGATORY_H_
+
+#cmakedefine01 TRITON_DALI_SKIP_DOWNLOAD
```
Comment (mzient) [inline src/utils/purgatory.h.in:26]: How about:
```suggestion
#cmakedefine01 TRITON_DALI_SKIP_DOWNLOAD
#cmakedefine CONDA_ENVIRONMENT_PATH "@CONDA_ENVIRONMENT@"
```
and removing `CONDA_ENVIRONMENT` from the compiler's command line?

**Review by mzient — DISMISSED** [review]
I still think that the intermediate constants are not really necessary, but I will not block the PR over it.

**Review by awolant — DISMISSED** [review]

**Comment by szalpal** [comment]
Merging the PR. It has been previously approved, but I had to fix a build error. Nothing significant changed between the approve and merge.

**Pull request merged**